llm/ext_server.go

```diff
@@ -0,0 +1,386 @@
+	buffer := &bytes.Buffer{}
+	enc := json.NewEncoder(buffer)
+	enc.SetEscapeHTML(false)
+
+	if err := enc.Encode(request); err != nil {
+		return fmt.Errorf("failed to marshal data: %v", err)
+	}
+
+	req := C.CString(buffer.String())
+	defer C.free(unsafe.Pointer(req))
+
+	errCh := make(chan error)
+	go func() {
+		errCh <- errWrap(C.llama_server_completion(req, (C.next_result)(unsafe.Pointer(C.nextResult_cgo)), user_data))
+	}()
+
+	select {
+	case <-ctx.Done():
+		deregisterCallback(user_data)
+	case err = <-errCh:
+	}
+	return err
+}
+
+func (llm *llamaExtServer) Encode(ctx context.Context, prompt string) ([]int, error) {
+	data, err := json.Marshal(TokenizeRequest{Content: prompt})
+	if err != nil {
+		return nil, fmt.Errorf("marshaling encode data: %w", err)
+	}
+	req := C.CString(string(data))
+	defer C.free(unsafe.Pointer(req))
+	var resp C.ext_server_resp
+	err = errWrap(C.llama_server_tokenize(req, &resp))
+	if resp.json_resp != nil {
+		defer C.free(unsafe.Pointer(resp.json_resp))
+	}
+
+	var encoded TokenizeResponse
+	if err2 := json.Unmarshal([]byte(C.GoString(resp.json_resp)), &encoded); err2 != nil {
+		return nil, fmt.Errorf("unmarshal encode response: %w", err2)
+	}
+
+	return encoded.Tokens, err
+}
+
+func (llm *llamaExtServer) Decode(ctx context.Context, tokens []int) (string, error) {
+	if len(tokens) == 0 {
+		return "", nil
+	}
+	data, err := json.Marshal(DetokenizeRequest{Tokens: tokens})
+	if err != nil {
+		return "", fmt.Errorf("marshaling decode data: %w", err)
+	}
+
+	req := C.CString(string(data))
+	defer C.free(unsafe.Pointer(req))
+	var resp C.ext_server_resp
+	err = errWrap(C.llama_server_detokenize(req, &resp))
+	if resp.json_resp != nil {
+		defer C.free(unsafe.Pointer(resp.json_resp))
+	}
+
+	var decoded DetokenizeResponse
+	if err2 := json.Unmarshal([]byte(C.GoString(resp.json_resp)), &decoded); err2 != nil {
+		return "", fmt.Errorf("unmarshal encode response: %w", err2)
+	}
+
+	return decoded.Content, err
+}
+
+func (llm *llamaExtServer) Embedding(ctx context.Context, input string) ([]float64, error) {
+	data, err := json.Marshal(TokenizeRequest{Content: input})
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling embed data: %w", err)
+	}
+
+	req := C.CString(string(data))
+	defer C.free(unsafe.Pointer(req))
+	var resp C.ext_server_resp
+	err = errWrap(C.llama_server_embedding(req, &resp))
+	if resp.json_resp != nil {
+		defer C.free(unsafe.Pointer(resp.json_resp))
+	}
+
+	var embedding EmbeddingResponse
+	if err := json.Unmarshal([]byte(C.GoString(resp.json_resp)), &embedding); err != nil {
+		return nil, fmt.Errorf("unmarshal tokenize response: %w", err)
+	}
+
+	return embedding.Embedding, nil
+}
+
+func (llm *llamaExtServer) Ping(ctx context.Context) error {
+	// TODO - consider some mechanism to check if the main loop and llama.cpp are in a good state
+	return nil
+}
+
+func (llm *llamaExtServer) Close() {
+	// TODO - do some shutdown stuff....
```

Comment:
The main thing to close here before was the external subprocess, so that shouldn't be a problem now. Just releasing whatever memory 👍